poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+ *   point provided at index bi of the promise. This verifies that the share
+ *   was actually intended for the guardian.
+ *
+ *   2. The provided share must fail to pass pubPoly.Check. This ensures that
+ *   the share is actually corrupted and the guardian is not just lying.
+ */
+
+type PromiseShare struct {
+
+	// The index of the share
+	i int
+	
+	// The actual share from the guardian
+	share abstract.Secret
+	
+	// The Diffie-Hellman key between guardian i and the promiser.
+	diffieKey abstract.Point
+}
+
+
+/* Initializes a new PromiseShare
+ *
+ * Arguments
+ *    i   = the index of the Promise share the guardian is revealing
+ *    s   = the share being revealed
+ *    d   = the Diffie-Hellman key between the guardian and promiser
+ *
+ * Returns
+ *   An initialized PromiseShare
+ */
+func (p *PromiseShare) Init(i int, s abstract.Secret, d abstract.Point) *PromiseShare {
+	p.i         = i
+	p.share     = s
+	p.diffieKey = d
+	return p
+}
+
+// Tests whether two promise secrets are equal.
+func (p *PromiseShare) Equal(p2 *PromiseShare) bool {
+	return p.i == p2.i && p.share.Equal(p2.share) &&
+	       p.diffieKey.Equal(p2.diffieKey)
+}
+
+/* Promise objects are mechanism by which servers can promise that a certain
+ * private key or abstract.Secret can be recomputed by other servers in case
+ * the original server goes offline.
+ *
+ * The Promise struct handles the logic of creating private shares, splitting
+ * these shares up for a given number of servers to act as guardians, verifying
+ * promises, and providing proof that guardians have indeed taken out approved
+ * of backing up the promse.
+ *
+ * Terms:
+ *   promiser = the server making the promise. The own who owns this object.
+ *   promised = recipients of the promise (aka the clients of the promiser)
+ *   guardian = another server who receives a share of the promise and can help
+ *              reconstruct it.
+ *
+ * Development note: The guardians, secrets, and signatures arrays should
+ *                   remain synchronized. In other words, the guardians[i],
+ *                   secrets[i], and signatures[i] should all refer to the same
+ *                   server.
+ */
+type Promise struct {
+
+	// The id of the promise. In the format:
+	//   PromiserPublicKey.String() + TimeOfCreation + RandomNumber
+	id string
+
+	// The cryptographic group to use for the private shares.
+	shareSuite abstract.Suite
+	
+	// The minimum number of shares needed to reconstruct the secret.
+	t int
+	
+	// The minimum number of shares needed before the policy can become
+	// active. t <= r
+	r int
+	
+	// The total number of shares to send.
+	n int
+	
+	// The public key of the promiser.
+	pubKey abstract.Point
+	
+	// The public polynomial that is used to verify that a secret share
+	// given did indeed come from the appropriate private key.
+	pubPoly *poly.PubPoly
+	
+	// Primarily for the promised, the number of shares that have been 
+	// revealed so far.
+	numShares int
```

Comment:
Looks like you're mixing "promise state" and "reconstruction state" in this struct - please separate the two.  Any specifically reconstruction-related state you need in the reconstruction process should be kept elsewhere.